extensions/typescript-language-features/src/languageFeatures/copilotRelated.ts

```diff
@@ -0,0 +1,76 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { nulToken } from '../utils/cancellation';
+import { jsTsLanguageModes, isSupportedLanguageMode } from '../configuration/languageIds';
+import { DocumentSelector } from '../configuration/documentSelector';
+import { API } from '../tsServer/api';
+import type * as Proto from '../tsServer/protocol/protocol';
+import { ITypeScriptServiceClient } from '../typescriptService';
+import { conditionalRegistration, requireMinVersion } from './util/dependentRegistration';
+
+const minVersion = API.v570;
+const dummyDisposable = new vscode.Disposable(() => { });
+
+export function register(
+	_selector: DocumentSelector,
+	client: ITypeScriptServiceClient,
+) {
+	return conditionalRegistration([
+		requireMinVersion(client, minVersion),
+	], () => {
+		const ext = vscode.extensions.getExtension('github.copilot');
+		if (!ext) {
+			return dummyDisposable;
+		}
+		ext.activate().then(() => {
+			const relatedAPI = ext.exports as {
+				registerRelatedFilesProvider(
+					providerId: { extensionId: string; languageId: string },
+					callback: (uri: vscode.Uri) => Promise<{ entries: vscode.Uri[]; traits?: { name: string; value: string }[] }>
+				): void;
+			} | undefined;
+			if (relatedAPI?.registerRelatedFilesProvider) {
+				for (const languageId of jsTsLanguageModes) {
+					const id = {
+						extensionId: 'vscode.typescript-language-features',
+						languageId
+					};
+					relatedAPI.registerRelatedFilesProvider(id, async uri => {
+						let document;
+						try {
+							document = await vscode.workspace.openTextDocument(uri);
+						} catch {
+							if (!vscode.window.activeTextEditor) {
+								vscode.window.showErrorMessage(vscode.l10n.t("Related files provider failed. No active text editor."));
+								return { entries: [] };
+							}
+							// something is REALLY wrong if you can't open the active text editor's document, so don't catch that
+							document = await vscode.workspace.openTextDocument(vscode.window.activeTextEditor.document.uri);
+						}
+
+						if (!isSupportedLanguageMode(document)) {
+							vscode.window.showErrorMessage(vscode.l10n.t("Related files provider failed. Copilot requested file with unsupported language mode."));
+							return { entries: [] };
+						}
+
+						const file = client.toOpenTsFilePath(document);
+						if (!file) {
+							return { entries: [] };
+						}
+						// @ts-expect-error until ts5.7
+						const response = await client.execute('copilotRelated', { file, }, nulToken) as Proto.CopilotRelatedResponse;
+						if (response.type !== 'response' || !response.body) {
+							return { entries: [] };
+						}
+						return { entries: response.body.relatedFiles.map(f => client.toResource(f)), traits: [] };
+					});
+				}
+			}
+		});
+		return dummyDisposable;
```

Comment:
Does `relatedAPI.registerRelatedFilesPro` return a disposable or is there a way of unregistering?

Reply:
No, but that's a good idea. I'll see what I can do there at the same time that I figure out cancellation.

Reply:
I wrote the code on the copilot side and got it turned into a PR, but I'll merge this as-is if that's OK.
(Unless the copilot PR gets merged very fast, in which case I'll add it to this PR.)

Reply:
The copilot PR is merged. I pushed a comment that uses the added Disposable and CancellationToken support.